runtime/src/dispatch.rs

```diff
@@ -119,65 +145,172 @@ impl<F, A> Dispatcher<F, A> {
-            None => maybe_into_block((self.func)(rt)?),
+            None => maybe_into_block((self.func)(rt)?, CBOR),
             Some(_) => Err(ActorError::illegal_argument("method expects no arguments".into())),
         }
     }
 }
 
-impl<'de, F, A, R, RT> Dispatch<'de, RT> for Dispatcher<F, (A,)>
+impl<F, R, RT, const CODEC: u64> Dispatch<RT> for Dispatcher<F, (WithCodec<R, CODEC>,)>
+where
+    F: FnOnce(&RT) -> Result<WithCodec<R, CODEC>, ActorError>,
+    R: Serialize,
+{
+    fn call(
+        self,
+        rt: &RT,
+        _method: u64,
+        args: Option<IpldBlock>,
+    ) -> Result<Option<IpldBlock>, ActorError> {
+        match args {
+            None => maybe_into_block((self.func)(rt)?.0, CODEC),
+            Some(_) => Err(ActorError::illegal_argument("method expects arguments".into())),
+        }
+    }
+}
+
+impl<F, A, R, RT> Dispatch<RT> for Dispatcher<F, (A, R)>
 where
     F: FnOnce(&RT, A) -> Result<R, ActorError>,
-    A: Deserialize<'de>,
+    A: DeserializeOwned,
     R: Serialize,
 {
-    fn call(self, rt: &RT, args: &'de Option<IpldBlock>) -> Result<Option<IpldBlock>, ActorError> {
+    fn call(
+        self,
+        rt: &RT,
+        _method: u64,
+        args: Option<IpldBlock>,
+    ) -> Result<Option<IpldBlock>, ActorError> {
         match args {
             None => Err(ActorError::illegal_argument("method expects arguments".into())),
-            Some(arg) => maybe_into_block((self.func)(rt, arg.deserialize()?)?),
+            Some(arg) => maybe_into_block((self.func)(rt, arg.deserialize()?)?, CBOR),
+        }
+    }
+}
+
+impl<F, A, R, RT, const CODEC: u64> Dispatch<RT> for Dispatcher<F, (WithCodec<A, CODEC>, R)>
+where
+    F: FnOnce(&RT, WithCodec<A, CODEC>) -> Result<R, ActorError>,
+    A: DeserializeOwned,
+    R: Serialize,
+{
+    fn call(
+        self,
+        rt: &RT,
+        _method: u64,
+        args: Option<IpldBlock>,
+    ) -> Result<Option<IpldBlock>, ActorError> {
+        match args {
+            None => Err(ActorError::illegal_argument("method expects arguments".into())),
+            Some(arg) if arg.codec != CODEC => {
+                Err(ActorError::illegal_argument("method expects arguments".into()))
+            }
+            Some(arg) => maybe_into_block((self.func)(rt, WithCodec(arg.deserialize()?))?, CBOR),
+        }
+    }
+}
+
+impl<F, A, R, RT, const CODEC: u64> Dispatch<RT> for Dispatcher<F, (A, WithCodec<R, CODEC>)>
+where
+    F: FnOnce(&RT, A) -> Result<WithCodec<R, CODEC>, ActorError>,
+    A: DeserializeOwned,
+    R: Serialize,
+{
+    fn call(
+        self,
+        rt: &RT,
+        _method: u64,
+        args: Option<IpldBlock>,
+    ) -> Result<Option<IpldBlock>, ActorError> {
+        match args {
+            None => Err(ActorError::illegal_argument("method expects arguments".into())),
+            Some(arg) => maybe_into_block((self.func)(rt, arg.deserialize()?)?.0, CODEC),
+        }
+    }
+}
+
+impl<F, A, R, RT, const CODEC: u64> Dispatch<RT>
```

Comment:
If this is intended to be the "flexible" method in which you can specify codecs for both the argument and result, should we have 2 separate CODEC constants?

(IIUC it's not used rn anyway since we don't have any such methods, but if we're including for completeness, may as well do the most flex thing?)

Reply:
Yeah, well, that's what I _meant_ to do...